cmd/config.go

```diff
@@ -151,51 +134,146 @@ func NewConfigSys() *ConfigSys {
+	}
+	version, err := configGetVersion(b)
+	if err != nil {
+		return nil, err
+	}
+	if version == serverConfigVersion {
+		// Already latest version, nothing more to be done.
+		return b, nil
+	}
+	b, err = migrateConfig(b)
+	if err != nil {
+		return nil, err
+	}
+	version, err = configGetVersion(b)
+	if err != nil {
+		return nil, err
+	}
+	if version != serverConfigVersion {
+		return nil, errIncorrectConfigVersion
+	}
+	// Save the migrated file in an indented fashion.
+	config := &serverConfig{}
+	if err = json.Unmarshal(b, config); err != nil {
+		return nil, err
+	}
+	if err = saveServerConfigIndent(context.Background(), objAPI, config); err != nil {
+		return nil, err
+	}
+	return b, nil
+}
+
 // Initialize and load config from remote etcd or local config directory
 func initConfig(objAPI ObjectLayer) error {
 	if objAPI == nil {
 		return errServerNotInitialized
 	}
 
-	configFile := path.Join(minioConfigPrefix, minioConfigFile)
-
-	if globalEtcdClient != nil {
-		if err := checkConfigEtcd(context.Background(), globalEtcdClient, getConfigFile()); err != nil {
-			if err == errConfigNotFound {
-				// Migrates all configs at old location.
-				if err = migrateConfig(); err != nil {
-					return err
-				}
-				// Migrates etcd ${HOME}/.minio/config.json to '/config/config.json'
-				if err = migrateConfigToMinioSys(objAPI); err != nil {
-					return err
-				}
-			} else {
-				return err
-			}
-		}
-
-		// Watch config for changes and reloads them.
-		go watchConfigEtcd(objAPI, configFile, loadConfig)
+	// Purge all configs with version '1',
+	// this is a special case since version '1' used
+	// to be a filename 'fsUsers.json' not 'config.json'.
+	if err := purgeV1(); err != nil {
+		return err
+	}
 
-	} else {
-		if isFile(getConfigFile()) {
-			if err := migrateConfig(); err != nil {
-				return err
-			}
+	configBytes, err := readServerConfig(context.Background(), objAPI)
+	if err != nil {
+		if err != errConfigNotFound {
+			return err
 		}
-		// Migrates ${HOME}/.minio/config.json or config.json.deprecated
-		// to '<export_path>/.minio.sys/config/config.json'
-		// ignore if the file doesn't exist.
-		if err := migrateConfigToMinioSys(objAPI); err != nil {
+		configBytes, err = configInitBackend(objAPI)
+		if err != nil {
 			return err
 		}
-
-		// Migrates backend '<export_path>/.minio.sys/config/config.json' to latest version.
-		if err := migrateMinioSysConfig(objAPI); err != nil {
+	}
+	version, err := configGetVersion(configBytes)
+	if err != nil {
+		return err
+	}
+	if version != serverConfigVersion {
+		configBytes, err = configMigrateBackend(objAPI)
+		if err != nil {
 			return err
 		}
 	}
 
+	// Rename config.json to config.json.deprecated
+	if isFile(getConfigFile()) {
+		os.Rename(getConfigFile(), getConfigFile()+".deprecated")
+	}
+
+	configFile := path.Join(minioConfigPrefix, minioConfigFile)
+
+	// Watch config for changes and reloads them in-memory.
+	go watchConfigEtcd(objAPI, configFile, loadConfig)
```

Comment:
This should always be inside `if globalEtcdClient != nil` block, it will crash if etcd is not set.